[RISC-V] Update ABI doc (#107286)
dotnet/runtime: docs/design/coreclr/botr/clr-abi.md

> Managed varargs are not supported in .NET Core.

Reviewer: ```suggestion
Managed varargs are not supported on Windows only.
```
The doc was not updated when managed varargs support was added on Windows in .NET Core 3.

Reviewer 2: Correct: *are supported*
Also clarify: is it supported on Windows ARM64?

> Managed/native varargs are supported on Windows only.

Reviewer: ```suggestion
Managed/native varargs are [supported on Windows only](https://github.com/dotnet/runtime/issues/82081).
```

> Managed/native varargs are supported on Windows only.

Reviewer: ```suggestion
Managed/native varargs are supported on Windows only. Support for managed/native varargs on non-Windows platforms is tracked by [this issue](https://github.com/dotnet/runtime/issues/82081).
```

> Managed/native varargs are currently not supported on RISC-V.

Reviewer: ```suggestion
Managed/native varargs are supported on Windows only.
```

Reviewer: There is no RISC-V version of Windows, so no RISC-V support falls out of that.

Author: Done

The AMD64 native calling conventions (Windows 64 and System V) require return buffer address to be returned by callee in RAX. JIT also follows this rule.

## RISC-V only: structs passed/returned according to hardware floating-point calling convention

> Passing/returning structs according to hardware floating-point calling convention like native is currently supported only up to 16 bytes, ones larger than that differ from the standard ABI and are passed/returned according to integer calling convention (by implicit reference).

Reviewer: What's the reason for this difference?

Author: TLDR, it's a bit of work for a nice-to-have and we're currently weeding out bugs and instabilities which are higher priority.

Supporting FP structs > 16 bytes entails we no longer can tell whether an argument is passed by implicit reference (or needs a return buffer) by looking at the struct size alone, we would have to:
1. calculate FpStructInRegistersInfo to tell if it's eligible for passing according to floating-point calling convention (can be mitigated by caching the classification like on System-V),
2. (for arguments only) even if 1. returns it's a legit FP struct, we'd still have to see if we have enough free registers to fit all fields, otherwise the struct is passed according to integer calling convention, which does cap the size of a struct passed by value to 16 bytes, so it can still be passed by reference.

My PoC in #101796 kind of supports that but at the cost of littering the code with #ifdef's, AFAIK no other architecture requires knowing register availability to tell if a parameter is passed by reference, including LoongArch with which we share a lot of ABI code. And ELT profiler currently has static scratch space for struct reconstruction, it would have to be precalculated. I also thought shuffle thunks would need work but knowing now https://github.com/dotnet/runtime/pull/107282#discussion_r1742784766 that's not the case.

And I still would have to figure out what to do about this, being static it doesn't have the context which registers are free:
https://github.com/dotnet/runtime/blob/57a4b04303ebb5a9418814d5d2778b08150a507b/src/coreclr/vm/callingconvention.h#L536-L561

Reviewer: Yeah, the RISCV calling convention is super complicated compared to everything else we have.

I think it would be fine to omit these complicated details from managed calling convention, and deal with these complicated cases in managed/unmanaged interop only.

Reviewer: What happens when one tries to use struct that is passed using floating point convention that is more than 16 bytes with managed/native interop?

If it does not work, it should be tracked by a bug. It needs to either throw an exception or work correctly.

Author: > Yeah, the RISCV calling convention is super complicated compared to everything else we have.

Maybe I'm just used to RISC-V but it doesn't feel more complicated than System-V, ARM64, or their .NET interops. Any calling convention has quirks and padding is a common source of them as noticed in #104098, #88238, and probably many more.

> I think it would be fine to omit these complicated details from managed calling convention, and deal with these complicated cases in managed/unmanaged interop only.

I'm thinking whether we need it in interops at all, it's a corner-case achievable only via extra padding and the user has an easy way out of explicitly passing by `ref`.

> What happens when one tries to use struct that is passed using floating point convention that is more than 16 bytes with managed/native interop?
>
> If it does not work, it should be tracked by a bug. It needs to either throw an exception or work correctly.

#107386

Small primitive arguments have undefined upper bits. This can be different from the standard calling conventions that may require normalization (e.g. on ARM32 and Apple ARM64).

> On RISC-V small primitive arguments are extended according to standard calling conventions.

Reviewer: The managed calling convention leaves the upper bits of small primitive arguments undefined across all architectures. It does not match the native calling convention wrt. passing small types on number of other architectures.

Is there a value for RISC-V to be special on this point?

Author: It's natural for RISC-V to have small primitives extended, any 32-bit instruction leaves the register extended so it's cheaper to require that of the caller (it's free in most cases) than to do 2 shifts for each argument every time in the callee, which we couldn't optimize out otherwise. Some care was already taken to properly extend small integers, e.g. in https://github.com/dotnet/runtime/pull/93665#discussion_r1365353888 so we're almost there anyway.

Is there a benefit of unifying it given each architecture strives to follow their own platform calling conventions?

Reviewer: Ok, sounds reasonable to me. It is another one of those things that make RISCV different.
